keps/sig-network/20191220-apiserver-url-envvar.md

```diff
@@ -0,0 +1,294 @@
+    - [Story 1](#story-1)
+    - [Story 2](#story-2)
+  - [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+    - [Examples](#examples)
+      - [Alpha -&gt; Beta Graduation](#alpha---beta-graduation)
+      - [Beta -&gt; GA Graduation](#beta---ga-graduation)
+      - [Removing a deprecated flag](#removing-a-deprecated-flag)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+Currently in-cluster API clients discover the apiserver endpoint using
+the `KUBERNETES_SERVICE_HOST` and `KUBERNETES_SERVICE_PORT`
+environment variables.  This KEP introduces a new
+`KUBERNETES_SERVICE_URL` environment variable and proposes deprecating
+the older environment variables.
+
+## Motivation
+
+`KUBERNETES_SERVICE_HOST` and `KUBERNETES_SERVICE_PORT` cannot be
+naively concatenated into a URL when `KUBERNETES_SERVICE_HOST` is an
+IPv6 literal address.  As introduced by [RFC 2732], IPv6 literal
+addresses must be escaped with square brackets when forming a URL to
+avoid ambiguity with URL port numbers - for example
+`http://[1080::8:800:200C:417A]:6443`.  This is unlike IPv4 literals
+or hostnames (even when hostnames resolve to IPv6 addresses), which
+cannot themselves contain colon characters.
+
+[RFC 2732]: https://tools.ietf.org/html/rfc2732
+
+Concatenating the `KUBERNETES_SERVICE_HOST` and `..._PORT`
+_incorrectly_ is a common error.  See for example the [java], [ruby],
+[haskell], and [javascript] clients.  This is particularly difficult
+in shell scripts, where there is no helper library available.
+
+[java]: kubernetes-client/java#839
+[ruby]: kubernetes-client/ruby#51
+[haskell]: kubernetes-client/haskell#68
+[javascript]: kubernetes-client/javascript#380
+
+Secondarily, the general `FOO_SERVICE_HOST` and `FOO_SERVICE_PORT`
+environment variables are a docker-compose-era legacy that works pooly
+in Kubernetes, and we would like to [remove support for them one
+day][servicevarskep].  Replacing the in-cluster Kubernetes API
+discovery logic removes the last "need" for this family of environment
+variables.
+
+[servicevarskep]: /keps/sig-apps/0028-20180925-optional-service-environment-variables.md
+
+### Goals
+
+- Provide a pointer to the in-cluster API server endpoint that "just
+  works" for IPv4, IPv6 and hostname forms.
+
+### Non-Goals
+
+- Actually remove `FOO_SERVICE_HOST`/`FOO_SERVICE_PORT` variables.
+  This is covered by [other KEPs][servicevarskep].
+
+## Proposal
+
+1. Unconditionally inject a new environment variable into pods called
+   `KUBERNETES_SERVICE_URL`.  The value would be the correctly-escaped
```

Comment:
Do we want to change the name at all while we're changing things? eg, `KUBERNETES_APISERVER_URL`? (Or just `KUBERNETES_APISERVER`).

Reply:
Longer is less likely to collide accidentally.

Reply:
Currently `KUBERNETES_SERVICE_HOST` is not populated if the [`kubernetes` service does not have a ClusterIP](https://github.com/kubernetes/kubernetes/blob/2a7225ff0e90e76f634b29d3a35a1ae3f0631bb7/pkg/kubelet/envvars/envvars.go#L42). Although it does not necessarily need to be spelled out in this KEP, it would be very useful to allow the `KUBERNETES_SERVICE_URL` hostname to be retrieved from an ExternalName type service. Which is [very close to being implemented here](https://github.com/kubernetes/kubernetes/blob/2a7225ff0e90e76f634b29d3a35a1ae3f0631bb7/pkg/kubelet/envvars/envvars.go#L66), but short-circuited in the logic inside `FromServices` above.

Reply:
Currently `KUBERNETES_SERVICE_HOST` is only populated if the `kubernetes` service [has a ClusterIP](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/envvars/envvars.go#L46). Although it may not need to be spelled out in this KEP, `KUBERNETES_SERVICE_URL` should also populate its hostname component by looking at the hostname of `kubernetes` if it is an `ExternalName` type service.

Reply:
The binary name doesn't matter - it is the kubernetes API server.

I could live with  KUBERNETES_API_URL - good enough?